### DISCUSSION TITLE
Where is TLS_method defined?

### DISCUSSION THREAD

**Discussion by RealTrisT** [openssl/openssl#23228] [Q&A] [answered]
There is a forward definition in ssl.h

```#define SSLv23_method           TLS_method
#define SSLv23_server_method    TLS_server_method
#define SSLv23_client_method    TLS_client_method

/* Negotiate highest available SSL/TLS version */
__owur const SSL_METHOD *TLS_method(void);
__owur const SSL_METHOD *TLS_server_method(void);
__owur const SSL_METHOD *TLS_client_method(void);
```

But I cannot for the life of me find the actual implementation.

Also, is there a place that describes the general layout of the code? I spent a while looking for BIO code, only to find it's in the crypto directory which I never would have guessed. The `bss_` prefix is also confusing. There seem to generally be no comments on the beginning of files, and no notes in the headers as to where the implementation is.

Thank you.

**Comment by mattcaswell** [comment, marked as answer]
It's implemented here using a macro:

https://github.com/openssl/openssl/blob/8d8c1600aed80b30e6dc00cbf1a9559feabec8ca/ssl/methods.c#L19-L22

The macros is defined here:

https://github.com/openssl/openssl/blob/8d8c1600aed80b30e6dc00cbf1a9559feabec8ca/ssl/ssl_local.h#L2276-L2315

> Also, is there a place that describes the general layout of the code? I spent a while looking for BIO code, only to find it's in the crypto directory which I never would have guessed. The bss_ prefix is also confusing. There seem to generally be no comments on the beginning of files, and no notes in the headers as to where the implementation is.

Not really, although there is some user focused documentation describing some of the services that libcrypto and libssl supply which might have given the hint that the BIO code is a libcrypto thing. It also useful background reading if you are trying to explore the code:

https://www.openssl.org/docs/manmaster/man7/ossl-guide-libraries-introduction.html

Also see:

https://www.openssl.org/docs/manmaster/man7/ossl-guide-libcrypto-introduction.html
https://www.openssl.org/docs/manmaster/man7/ossl-guide-libssl-introduction.html

**Reply by RealTrisT** [reply]
What is the reason for TLS_ANY_VERSION to be using `TLSv1_2_enc_data` when `TLSv1_3_enc_data` exists? Is it overwritten after the version is negotiated in the handshake?

**Reply by mattcaswell** [reply]
As soon as an actual TLS version is selected the method gets swapped out to a "real" method. The "enc_data" is mostly irrelevant prior to that point.
